File: leetcode/tree/617.合并二叉树.ts
/*
 * @lc app=leetcode.cn id=617 lang=typescript
 *
 * [617] 合并二叉树
 *
 * https://leetcode-cn.com/problems/merge-two-binary-trees/description/
 *
 * algorithms
 * Easy (77.23%)
 * Likes:    486
 * Dislikes: 0
 * Total Accepted:    76.3K
 * Total Submissions: 98.3K
 * Testcase Example:  '[1,3,2,5]\n[2,1,3,null,4,null,7]'
 *
 * 给定两个二叉树，想象当你将它们中的一个覆盖到另一个上时，两个二叉树的一些节点便会重叠。
 *
 * 你需要将他们合并为一个新的二叉树。合并的规则是如果两个节点重叠，那么将他们的值相加作为节点合并后的新值，否则不为 NULL
 * 的节点将直接作为新二叉树的节点。
 *
 * 示例 1:
 *
 *
 * 输入:
 * Tree 1                     Tree 2
 * ⁠         1                         2
 * ⁠        / \                       / \
 * ⁠       3   2                     1   3
 * ⁠      /                           \   \
 * ⁠     5                             4   7
 * 输出:
 * 合并后的树:
 * 3
 * / \
 * 4   5
 * / \   \
 * 5   4   7
 *
 *
 * 注意: 合并必须从两个树的根节点开始。
 *
 */

export {};

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// @lc code=start
// dfs
var mergeTrees = function (
  t1: TreeNode | null,
  t2: TreeNode | null
): TreeNode | null {
  if (!t1 && !t2) return null;
  if (!t1 && t2) return t2;
  if (t1 && !t2) return t1;
  const root = new TreeNode(t1!.val + t2!.val);
  root.left = mergeTrees(t1!.left, t2!.left);
  root.right = mergeTrees(t1!.right, t2!.right);
  return root;
};

// bfs
var mergeTrees = function (
  t1: TreeNode | null,
  t2: TreeNode | null
): TreeNode | null {
  if (!t1 && !t2) return null;
  if (!t1 && t2) return t2;
  if (t1 && !t2) return t1;

  const queue: TreeNode[] = [];
  const queue1: TreeNode[] = [];
  const queue2: TreeNode[] = [];
  const t = new TreeNode(t1!.val + t2!.val);
  queue.push(t);
  queue1.push(t1!);
  queue2.push(t2!);

  while (queue1.length || queue2.length) {
    const node = queue.shift()!;
    const node1 = queue1.shift()!;
    const node2 = queue2.shift()!;

    const left1 = node1.left;
    const right1 = node1.right;
    const left2 = node2.left;
    const right2 = node2.right;

    if (left1 && left2) {
      const left = new TreeNode(left1.val + left2.val);
      node.left = left;
      queue.push(left);
      queue1.push(left1);
      queue2.push(left2);
    } else if (left1) {
      node.left = left1;
    } else if (left2) {
      node.left = left2;
    }

    if (right1 && right2) {
      const right = new TreeNode(right1.val + right2.val);
      node.right = right;
      queue.push(right);
      queue1.push(right1);
      queue2.push(right2);
    } else if (right1) {
      node.right = right1;
    } else if (right2) {
      node.right = right2;
    }
  }

  return t;
};
// @lc code=end

File: leetcode/tree/617.合并二叉树.go
/*
 * @lc app=leetcode.cn id=617 lang=golang
 *
 * [617] 合并二叉树
 *
 * https://leetcode.cn/problems/merge-two-binary-trees/description/
 *
 * algorithms
 * Easy (79.15%)
 * Likes:    1272
 * Dislikes: 0
 * Total Accepted:    425.4K
 * Total Submissions: 536.5K
 * Testcase Example:  '[1,3,2,5]\n[2,1,3,null,4,null,7]'
 *
 * 给你两棵二叉树： root1 和 root2 。
 *
 *
 * 想象一下，当你将其中一棵覆盖到另一棵之上时，两棵树上的一些节点将会重叠（而另一些不会）。你需要将这两棵树合并成一棵新二叉树。合并的规则是：如果两个节点重叠，那么将这两个节点的值相加作为合并后节点的新值；否则，不为
 * null 的节点将直接作为新二叉树的节点。
 *
 * 返回合并后的二叉树。
 *
 * 注意: 合并过程必须从两个树的根节点开始。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：root1 = [1,3,2,5], root2 = [2,1,3,null,4,null,7]
 * 输出：[3,4,5,5,4,null,7]
 *
 *
 * 示例 2：
 *
 *
 * 输入：root1 = [1], root2 = [1,2]
 * 输出：[2,2]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 两棵树中的节点数目在范围 [0, 2000] 内
 * -10^4 <= Node.val <= 10^4
 *
 *
 */

package leetcode

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
// dfs
// func mergeTrees(root1 *TreeNode, root2 *TreeNode) *TreeNode {
// 	if root1 == nil {
// 		return root2
// 	}
// 	if root2 == nil {
// 		return root1
// 	}
// 	root := &TreeNode{
// 		Val: root1.Val + root2.Val,
// 	}
// 	root.Left = mergeTrees(root1.Left, root2.Left)
// 	root.Right = mergeTrees(root1.Right, root2.Right)
// 	return root
// }

// bfs
func mergeTrees(root1 *TreeNode, root2 *TreeNode) *TreeNode {
	if root1 == nil {
		return root2
	}
	if root2 == nil {
		return root1
	}
	root := &TreeNode{
		Val: root1.Val + root2.Val,
	}
	queue := []*TreeNode{root}
	queue1 := []*TreeNode{root1}
	queue2 := []*TreeNode{root2}

	for len(queue1) > 0 || len(queue2) > 0 {
		node := queue[0]
		queue = queue[1:]
		node1 := queue1[0]
		queue1 = queue1[1:]
		node2 := queue2[0]
		queue2 = queue2[1:]

		left1 := node1.Left
		right1 := node1.Right
		left2 := node2.Left
		right2 := node2.Right

		if left1 != nil && left2 != nil {
			left := &TreeNode{
				Val: left1.Val + left2.Val,
			}
			node.Left = left
			queue = append(queue, left)
			queue1 = append(queue1, left1)
			queue2 = append(queue2, left2)
		} else if left1 != nil {
			node.Left = left1
		} else if left2 != nil {
			node.Left = left2
		}

		if right1 != nil && right2 != nil {
			right := &TreeNode{
				Val: right1.Val + right2.Val,
			}
			node.Right = right
			queue = append(queue, right)
			queue1 = append(queue1, right1)
			queue2 = append(queue2, right2)
		} else if right1 != nil {
			node.Right = right1
		} else if right2 != nil {
			node.Right = right2
		}
	}

	return root
}

// @lc code=end
